jetstream.go

```diff
@@ -0,0 +1,334 @@
+		return nil, ErrNotJSMessage
+	}
+
+	var err error
+
+	meta.stream = parts[2]
+	meta.consumer = parts[3]
+	meta.delivered, err = strconv.Atoi(parts[4])
+	if err != nil {
+		return nil, ErrNotJSMessage
+	}
+
+	meta.sSeq, err = strconv.Atoi(parts[5])
+	if err != nil {
+		return nil, ErrNotJSMessage
+	}
+
+	meta.cSeq, err = strconv.Atoi(parts[6])
+	if err != nil {
+		return nil, ErrNotJSMessage
+	}
+
+	tsi, err := strconv.Atoi(parts[7])
+	if err != nil {
+		return nil, ErrNotJSMessage
+	}
+	meta.ts = time.Unix(0, int64(tsi))
+
+	meta.parsed = true
+
+	return meta, nil
+}
+
+const jsStreamUnspecified = "not.set"
+
+type jsAcKOpts struct {
+	str string // stream to expect a ack from
+}
+
+type jsOpts struct {
+	timeout time.Duration
+	ctx     context.Context
+
+	ack jsAcKOpts
+}
+
+func newJsOpts() *jsOpts {
+	return &jsOpts{ack: jsAcKOpts{str: jsStreamUnspecified}}
+}
+
+func (j *jsOpts) context(dftl time.Duration) (context.Context, context.CancelFunc) {
+	if j.ctx == nil && j.timeout == 0 {
+		j.timeout = dftl
+	}
+
+	if j.ctx != nil {
+		return context.WithCancel(j.ctx)
+	}
+
+	return context.WithTimeout(context.Background(), j.timeout)
+}
+
+// AckOption configures the various JetStream message acknowledgement helpers
+type AckOption func(opts *jsOpts) error
+
+// PublishOption configures publishing messages
+type PublishOption func(opts *jsOpts) error
+
+// PublishExpectsStream waits for an ack after publishing and ensure it's from a specific stream, empty arguments waits for any valid acknowledgement
+func PublishExpectsStream(stream ...string) PublishOption {
+	return func(opts *jsOpts) error {
+		switch len(stream) {
+		case 0:
+			opts.ack.str = ""
+
+		case 1:
+			opts.ack.str = stream[0]
+			if !isValidJSName(opts.ack.str) {
+				return ErrInvalidStreamName
+			}
+
+		default:
+			return ErrInvalidStreamName
+		}
+
+		return nil
+	}
+}
+
+// PublishStreamTimeout sets the period of time to wait for JetStream to acknowledge receipt, defaults to JetStreamTimeout option
+func PublishStreamTimeout(t time.Duration) PublishOption {
+	return func(opts *jsOpts) error {
+		opts.timeout = t
+		return nil
+	}
+}
+
+// PublishCtx sets an interrupt context for waiting on a stream to reply
+func PublishCtx(ctx context.Context) PublishOption {
+	return func(opts *jsOpts) error {
+		opts.ctx = ctx
+		return nil
+	}
+}
+
+// AckWaitDuration waits for confirmation from the JetStream server
+func AckWaitDuration(d time.Duration) AckOption {
+	return func(opts *jsOpts) error {
+		opts.timeout = d
+		return nil
+	}
+}
+
+func (m *Msg) jsAck(body []byte, opts ...AckOption) error {
+	if m.Reply == "" {
+		return ErrMsgNoReply
+	}
+
+	if m == nil || m.Sub == nil {
+		return ErrMsgNotBound
+	}
+
+	m.Sub.mu.Lock()
+	nc := m.Sub.conn
+	m.Sub.mu.Unlock()
+
+	var err error
+
+	aopts := newJsOpts()
+	for _, f := range opts {
+		err = f(aopts)
+		if err != nil {
+			return err
+		}
+	}
+
+	if aopts.timeout == 0 {
```

Comment:
Wondering if this low level function should explicitly take an optional reply subject etc for ack of ack.

Reply:
not needed really since we are on m and have access to m.Reply as below? This does ack of ack if reply is set and options indicate we should wait for any reason